File: adivina.php
<?php

$secretNumber = null;
$attempts = null;

function setName()
{
    echo 'Ingrese su nombre: ';
    $name = trim(readline());

    if ($name === "") {
        echo 'No se admiten campos vacíos. Ingrese un nombre de usuario.' . "\n";
        return;
    }

    echo 'Bienvenido ' . $name . '.' . "\n";
    resetGame();
}

function resetGame()
{
    global $secretNumber, $attempts;
    $secretNumber = rand(1, 100);
    $attempts = 6;
    echo 'Tienes ' . $attempts . ' intentos restantes.' . "\n";
    echo "\n";
    checkNumber();
}

function checkNumber()
{
    global $secretNumber, $attempts;
    echo 'Ingrese un número del 1 al 100: ';
    $userNumber = trim(readline());

    if ($userNumber < 1 || $userNumber > 100) {
        echo 'Solo puedes introducir números del 1 al 100.' . "\n";
        echo "\n";
        return;
    }

    $attempts--;

    if ($userNumber < $secretNumber) {
        echo 'El número ingresado es menor. Inténtalo de nuevo.' . "\n";
        echo 'Tienes ' . $attempts . ' intentos restantes.' . "\n";
    } elseif ($userNumber > $secretNumber) {
        echo 'El número ingresado es mayor. Inténtalo de nuevo.' . "\n";
        echo 'Tienes ' . $attempts . ' intentos restantes.' . "\n";
    } else {
        echo '¡Felicidades, has adivinado!' . "\n";
        echo "\n";
        retry();
    }

    if ($attempts == 0 && $userNumber != $secretNumber) {
        echo 'Lo siento, has perdido. El número era ' . $secretNumber . "\n";
        echo "\n";
        retry();
    }
}

function retry()
{
    echo '¿Deseas jugar de nuevo? (S/N): ';
    $playAgain = trim(readline());

    if (strtolower($playAgain) == 's') {
        resetGame();
    } else {
        echo 'Gracias por jugar. ¡Hasta luego!' . "\n";
    }
}

setName();

File: adivina.py
import random

secretNumber = None
attempts = None

def setName():
    name = input('Ingrese su nombre: ')

    if name == "":
        print('No se admiten campos vacíos. Ingrese un nombre de usuario.')
        return

    print('Bienvenido ' + name + '.')
    resetGame()

def resetGame():
    global secretNumber, attempts
    secretNumber = random.randint(1, 100)
    attempts = 6
    print('Tienes', attempts, 'intentos restantes.')
    print('')
    checkNumber()

def checkNumber():
    global secretNumber, attempts
    userNumber = int(input('Ingrese un número del 1 al 100: '))

    if userNumber < 1 or userNumber > 100:
        print('Solo puedes introducir números del 1 al 100.')
        print('')
        return

    attempts -= 1

    if userNumber < secretNumber:
        print('El número ingresado es menor. Inténtalo de nuevo.')
        print('Tienes', attempts, 'intentos restantes.')
    elif userNumber > secretNumber:
        print('El número ingresado es mayor. Inténtalo de nuevo.')
        print('Tienes', attempts, 'intentos restantes.')
    else:
        print('¡Felicidades, has adivinado!')
        print('')
        retry()

    if attempts == 0 and userNumber != secretNumber:
        print('Lo siento, has perdido. El número era', secretNumber)
        print('')
        retry()

def retry():
    playAgain = input('¿Deseas jugar de nuevo? (S/N): ')

    if playAgain.lower() == 's':
        resetGame()
    else:
        print('Gracias por jugar. ¡Hasta luego!')


setName()
